platform/windows/os_windows.cpp

```diff
@@ -622,6 +622,74 @@ Vector<String> OS_Windows::get_video_adapter_driver_info() const {
 	return info;
 }
 
+bool OS_Windows::get_user_prefers_integrated_gpu() const {
+	// On Windows 10, the preferred GPU configured in Windows Settings is
+	// stored in the registry under the key
+	// `HKEY_CURRENT_USER\SOFTWARE\Microsoft\DirectX\UserGpuPreferences`
+	// with the name being the app ID or EXE path. The value is in the form of
+	// `GpuPreference=1;`, with the value being 1 for integrated GPU and 2
+	// for discrete GPU. On Windows 11, there may be more flags, separated
+	// by semicolons.
+
+	// If this is a packaged app, use the "application user model ID".
+	// Otherwise, use the EXE path.
+	WCHAR value_name[32768];
+	bool is_packaged = false;
+	{
+		HMODULE kernel32 = GetModuleHandleW(L"kernel32.dll");
+		if (kernel32) {
+			using GetCurrentApplicationUserModelIdPtr = LONG (*)(UINT32 *length, PWSTR id);
+			GetCurrentApplicationUserModelIdPtr GetCurrentApplicationUserModelId = (GetCurrentApplicationUserModelIdPtr)GetProcAddress(kernel32, "GetCurrentApplicationUserModelId");
+
+			if (GetCurrentApplicationUserModelId) {
+				UINT32 length = sizeof(value_name) / sizeof(value_name[0]);
+				LONG result = GetCurrentApplicationUserModelId(&length, value_name);
+				if (result == ERROR_SUCCESS) {
+					is_packaged = true;
+				} else if (result != APPMODEL_ERROR_NO_APPLICATION) {
+					return false;
+				}
+			}
+		}
+	}
+	if (!is_packaged && GetModuleFileNameW(nullptr, value_name, sizeof(value_name) / sizeof(value_name[0])) >= sizeof(value_name) / sizeof(value_name[0])) {
+		// Paths should never be longer than 32767, but just in case.
+		return false;
+	}
+
+	LPCWSTR subkey = L"SOFTWARE\\Microsoft\\DirectX\\UserGpuPreferences";
+	HKEY hkey = nullptr;
+	LSTATUS result = RegOpenKeyExW(HKEY_CURRENT_USER, subkey, 0, KEY_READ, &hkey);
+	if (result != ERROR_SUCCESS) {
+		return false;
+	}
+
+	DWORD size = 0;
+	result = RegGetValueW(hkey, nullptr, value_name, RRF_RT_REG_SZ, nullptr, nullptr, &size);
```

Comment:
Reg. key name size is limited to 16K chars, I wounder what would happen if you try to set GPU preferences for app with longer path (will it truncate it or won't set it at all).

Probably not relevant in practice, Godot won't handle long paths without https://github.com/godotengine/godot/pull/91902, and `CreateProcess` can't handle long current directory paths, so it's unlikely will ever be over 260.

Reply:
Interesting point to consider. Though things are really whacky even when I make the Godot exe path just 256 chars -- it crashes when using the NVIDIA GPU *and* explorer.exe; even if I set it to use the Intel GPU (OpenGL selection is handled by the system) it still crashed explorer.exe (though it did run the project manager).

I think I am comfortable with leaving the code like this. It is obviously not a supported scenario (unless MS decides to raise the limit of the value name, which I doubt would ever happen) and `RegOpenKeyExW` should just return an error and the function returns false.